module/zfs/zfs_znode.c

```diff
@@ -1049,6 +1075,11 @@ zfs_xvattr_set(znode_t *zp, xvattr_t *xvap, dmu_tx_t *tx)
 		    zp->z_pflags, tx);
 		XVA_SET_RTN(xvap, XAT_SPARSE);
 	}
+	if (XVA_ISSET_REQ(xvap, XAT_PROJINHERIT)) {
+		ZFS_ATTR_SET(zp, ZFS_PROJINHERIT, xoap->xoa_projinherit,
+		    zp->z_pflags, tx);
```

Comment:
@behlendorf @Nasf-Fan After reading the xfs_quota manpage a few times, I think that in XFS, the tree quota is not really enforced, i.e. you can end up with a tree where some files are not assigned to the project.  Hence the need for `xfs_quota project -c`.  `xfs_quota project -d` explicitly sets up such "broken" project trees.  This seems like a bug (or misdesign), rather than a feature.  In such a system, you might have a PROJINHERIT flag on the directory that means "if you didn't say otherwise, default to setting the project the same as the containing directory".  But you might also not have such a flag, and instead just make that behavior the default.

I think you're saying that the zfs directory tree project mechanism should behave the same as XFS.  IMO, it would be unfortunate to follow a poor design simply because it came first.  I think we can achieve a better design which is still "logically compatible" with the typical uses of xfs quotas (i.e. ignoring `xfs_quota project -d`).

I propose that PROJINHERIT should be enforced by the kernel (with the restrictions I mentioned: it can be set only after checking that all its (direct) entries have the matching PROJID, and directories have PROJINHERIT; and it can only be cleared if its parent does not have it set), and the `zfs project` commands would do the recursive traversal in userland to set the PROJID and PROJINHERIT such that the kernel's checks will succeed.  Thus we can check in O(1) time if a tree is all assigned the same project.  As you said, `zfs project -c` can check this, but unlike the XFS version it would not "reports nothing if the tree is correct, otherwise it reports the paths of inodes which do not have the project ID of the rest of the tree, or if the inode flag is not set", because these situations would not be possible.

`zfs project -s` would, from userland, do a depth-first traversal of the directory, setting the PROJID on all files and directories, and setting PROJINHERIT on all directories.  When setting PROJINHERIT, the kernel would check that all contained files and directories have the same PROJID, and all contained directories have PROJINHERIT already set.  So although setting PROJINHERIT is O(N entries), the overall `zfs project -s` operation is necessarily at least O(N entries), so the enforcement logic isn't making the time complexity any worse.

As a side note, how does `/etc/projects` work?  It it a shorthand so that you can specify a "project" on the CLI and it uses /etc/projects to map to a path? I don't really understand the need for it.  You could specify the directory path directly on the CLI, e.g. `zfs project -s <dir> <project>`